cachi2/core/package_managers/pip.py

```diff
@@ -1370,20 +1453,140 @@ def _split_hashes_from_options(cls, options: list[str]) -> tuple[list[str], list
+        verified: bool = False
+        try:
+            must_match_any_checksum(path, checksum_info)
+            verified = True
+        except PackageRejected:
+            path.unlink()
+            log.warning("Download '%s' was removed from the output directory", path.name)
+        return verified
+
+    if dpi:
+        if dpi.should_verify_checksums():
+            download_info["hash_verified"] = _hash_verify(dpi.path, dpi.checksums_to_verify)
+
+        if dpi.package_type == "sdist":
+            _check_metadata_in_sdist(dpi.path)
+
+    else:
+        if require_hashes or req.kind == "url":
+            hashes = req.hashes or [req.qualifiers.get("cachito_hash", "")]
+            download_info["hash_verified"] = _hash_verify(
+                download_info["path"], list(map(_to_checksum_info, hashes))
+            )
+
+    log.info(
+        "-- Successfully processed '%s' in path '%s'",
+        req.download_line,
+        download_info["path"].relative_to(pip_deps_dir.root),
+    )
+
+    return download_info
+
+
+def _process_pypi_req(
+    req: PipRequirement,
+    requirements_file: PipRequirementsFile,
+    require_hashes: bool,
+    pip_deps_dir: RootedPath,
+    allow_binary: bool,
+) -> list[dict[str, Any]]:
+    download_infos: list[dict[str, Any]] = []
+
+    def _process_req_wrapper(
+        DPIs: list[DistributionPackageInfo], log_string: str
+    ) -> list[dict[str, Any]]:
+        infos: list[dict[str, Any]] = []
+        for dpi in DPIs:
+            log.info(log_string, dpi.path.name)
+            infos.append(
+                _process_req(
+                    req, requirements_file, require_hashes, pip_deps_dir, dpi.download_info, dpi=dpi
+                )
+            )
+        return infos
+
+    artifacts: list[DistributionPackageInfo] = _process_package_distributions(
+        req, pip_deps_dir, allow_binary
+    )
+    artifacts_to_check: list[DistributionPackageInfo] = [a for a in artifacts if a.path.exists()]
+    artifacts_to_download: list[DistributionPackageInfo] = [
+        a for a in artifacts if not a.path.exists()
+    ]
+
+    files: dict[str, Union[str, PathLike[str]]] = {
+        dpi.url: dpi.path for dpi in artifacts_to_download
+    }
+    asyncio.run(async_download_files(files, get_config().concurrency_limit))
+
+    download_infos.extend(
+        _process_req_wrapper(artifacts_to_check, "-- Checking existing pypi artifact '%s'")
+    )
+    download_infos.extend(
+        _process_req_wrapper(artifacts_to_download, "-- Processing pypi artifact to download '%s'")
+    )
+
+    return download_infos
+
+
+def _process_vcs_req(
+    req: PipRequirement, pip_deps_dir: RootedPath, **kwargs: Any
+) -> dict[str, Any]:
+    log.info("-- Processing vcs repo '%s'", req.url)
+
+    return _process_req(
+        req,
+        pip_deps_dir=pip_deps_dir,
+        download_info=_download_vcs_package(req, pip_deps_dir),
+        **kwargs,
```

Comment:
Why only `_process_pypi_req` does not have `**kwargs` and `_process_vcs_req`, `_process_url_req` do?

Reply:
IIRC, `_process_pypi_req` doesn't need it, but I'll double-check

Reply:
because unlike `_process_pypi_req` both `_process_vcs_req` and `_process_url_req` don't make use of most of the arguments `_process_req` takes, so here `**kwargs` represents just that - **passthrough arguments**.